brainiak/matnorm_models/base.py

```diff
@@ -0,0 +1,229 @@
+            max_abs_current_grad = self.sess.run(max_abs_current_grad_op,
+                                                 feed_dict=feed_dict)
+
+            # check tolerances
+            if abs((current_loss - past_loss) /
+                    max(current_loss, past_loss)) < loss_tol:
+                print('loss tolerance reached on iter %i, %f, stopping' %
+                      (n+1, current_loss))
+                break
+            if max_abs_current_grad < grad_tol:
+                print('gradient tolerance reached on iter %i, %f, stopping' %
+                      (n+1, current_loss))
+                break
+            if (n+1) % step == 0:
+                print('iter %i, %f' % (n+1, current_loss))
+
+    def solve_det_marginal(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma + AQA')^{-1} X =\\
+             \Sigma^{-1} - \Sigma^{-1} A (Q^{-1} + A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma + AQA')| = \log|Q^{-1} + A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # cholesky of (Qinv + A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma_inv + tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Qinv + A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+        solve = tf.matmul(sigma.Sigma_inv -
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
+        return solve, logdet
+
+    def solve_det_conditional(self, x, sigma, A, Q):
+        """
+        Use matrix inversion lemma for the solve:
+        .. math::
+            (\Sigma - AQA')^{-1} X =\\
+             \Sigma^{-1} + \Sigma^{-1} A (Q^{-1} - A' \Sigma^{-1} A)^{-1} A' \Sigma^{-1}
+
+        Use matrix determinant lemma for determinant:
+        ..math::
+            \log|(\Sigma - AQA')| = \log|Q^{-1} - A' \Sigma^{-1} A| + \log|Q| + \log|\Sigma|
+        """
+
+        # (Qinv - A' Sigma^{-1} A)
+        i_qf_cholesky = tf.cholesky(Q.Sigma - tf.matmul(A,
+                                    sigma.Sigma_inv_x(A), transpose_a=True))
+
+        logdet = -Q.logdet + sigma.logdet +\
+            2 * tf.reduce_sum(tf.log(tf.matrix_diag_part(i_qf_cholesky)))
+
+        # A' Sigma^{-1}
+        Atrp_Sinv = tf.matmul(A, sigma.Sigma_inv, transpose_a=True)
+        # (Q - A' Sigma^{-1} A)^{-1} A' Sigma^{-1}
+        prod_term = tf.cholesky_solve(i_qf_cholesky, Atrp_Sinv)
+
+        solve = tf.matmul(sigma.Sigma_inv +
+                          sigma.Sigma_inv_x(tf.matmul(A, prod_term)), x)
```

Comment:
If x is a vector (or a small column size matrix, then you should consider multiplying by x first.
(Sinv + Sinv*(A*Prod))*X  = Sinv*(I + A*Prod)*x.
So you can call once to Sigma_inv_x()
Also you can consider the cholesky solve above  for x+A*Prod_x and you can compute first Prod_x =chol_solve(i_qf_chol, A'*sigma_inv_x(x))

Everything depends on the size of x and the matrices. 
A better way to do this is using [a conditional structure](https://www.tensorflow.org/versions/master/api_docs/python/control_flow_ops/control_flow_operations#cond).